src/svgp.jl

```diff
@@ -79,5 +79,47 @@ end
 
 inducing_points(f::ApproxPosteriorGP{<:SVGP}) = f.approx.fz.x
 
-_chol_cov(q::AbstractMvNormal) = cholesky(Symmetric(cov(q)))
-_chol_cov(q::MvNormal) = cholesky(q.Σ)
+## Pathwise posterior sampling
+
+struct PosteriorFunctionSamples
+    num_samples
+    prior  # prior(x) returns an approx prior sample
+    update  # update(x) computes the required pathwise update
+end
+
+(f::PosteriorFunctionSamples)(x) = f([x])
+(f::PosteriorFunctionSamples)(x::AbstractVector) = f.prior(x) + f.update(x)
+
+# TODO: docstrings
+function pathwise_sample(
+    rng::AbstractRNG,
+    f::ApproxPosteriorGP{<:SVGP},
+    prior_sample_function; # TODO: better name?
+    num_samples=1::Int,
+)
+    svgp = f.approx
+    z = svgp.fz.x
+    input_dims = length(z[1])  # TODO: what's the best way of getting this?
```

Comment:
I've been thinking about this. My thinking is that we need to introduce a new kernel type somewhere that wraps an existing kernel and gives us information about the dimension of input that we expect it to apply to -- probably we want to add this in KernelFunctions.

Something like
```julia
# Euclidean domain with D dimensions.
struct Euclidean
    D::Int
end

dim(domain::Euclidean) = domain.D

"""
    DomainKernel{Tkernel<:Kernel, Tdomain}

A kernel with additional information specifying the domain of the kernel.
This provides a complete specification of a kernel, whereas most kernels by themselves do
not provide the domain information necessary to construct e.g. Random Fourier Feature
approximations to themselves.
"""
struct DomainKernel{Tkernel<:Kernel, Tdomain} <: Kernel
    kernel::Tkernel
    domain::Tdomain
end

kernelmatrix(k::DomainKernel, x::AbstractVector) = kernelmatrix(k.kernel, x)

function kernelmatrix(k::DomainKernel, x::AbstractVector, x′::AbstractVector)
    return kernelmatrix(k.kernel, x, x′)
end

kernelmatrix_diag(k::DomainKernel, x::AbstractVector) = kernelmatrix_diag(k.kernel, x)
```

I've written some of my own RFF code as I'm using pathwise sampling in some of my PhD work -- the conclusion that I came to was that our kernels simply don't provide a sufficient amount of information to construct RFF approximations, and really we ought to be able to construct the approximation just given the kernel.

Equally, we don't want to force users to always provide dimension information, so we need something that doesn't break existing code, but allows users to provide enough information to construct the RFF approximation if that's what they want to do.

Reply:
Related issue: https://github.com/JuliaGaussianProcesses/KernelFunctions.jl/issues/16
Incidentally involving Alex as well :laughing:

Reply:
> @willtebbutt you argued yourself against having kernels defining their own domain your(past)self 😆 JuliaGaussianProcesses/KernelFunctions.jl#16 (comment)

Indeed I did. IIRC my main concern there was the effect that it would have on end-users -- I'm less concerned if we're able to provide a default domain that users don't see. (If that wasn't my main concern there, it's fair to say that it's my main concern now 😂 )